Show password login webview and prepare internal architecture

adalpari opened the pull request:
## Description

This PR is adding a new step for the Application Password login. In this new iteration we are showing to the user the authentication webview and intercepting the callback url. NOTE: the callback is yet not handled. That will be done in the next iteration: handling it and storing it.

## Testing instructions

1. Run the app and test BOTH login flow to be sure nothing is broken. They should work as always.
2. Open the project and, in `LoginSiteAddressFragment`, comment the XML RPC lines as showing in the following screenshots (231, 232, 340) and uncomment the REST ones (229, 343). _I could use a Feature Flag to control it, but I'm not sure it is worth it since I don't see them very accessible from the libs, so it would require more extra logic. But I accept suggestions :)_
<img width="954" alt="Screenshot 2025-05-02 at 16 26 54" src="https://github.com/user-attachments/assets/1edfd9ce-117d-40de-a504-8c5d6e570118" />
<img width="1029" alt="Screenshot 2025-05-02 at 16 26 40" src="https://github.com/user-attachments/assets/7d6e0e58-c05e-4559-8a37-57d848cd3e88" />
3. Run the app again. And login in a self-hosted site
- [ ] Verify you see a log like: `Found authorization for https://vanilla.wpmt.co URL: https://vanilla.wpmt.co/wp-admin/authorize-application.php?app_name=android-jetpack-client&success_url=jetpack%3A%2F%2Fwpcom-authorize`
- [ ] Verify you are redirected to the authentication webview
4. Approve the connection
- [ ] Verify you are redirected to the login screen (that's enough for now)

https://github.com/user-attachments/assets/eb1a0703-73d5-476f-b912-ef074824aa6e

<!-

-

Consider testing the following *yourself* before requesting a review:

- Compatibility with WordPress.com sites and self-hosted Jetpack sites
- Both portrait and landscape orientations
- Light and dark modes
- Dynamic font sizes: larger, smaller and bold text
- High contrast mode
- Screen reader experience (e.g., Talkback)
- Languages with large words or with letters/accents not frequently used in English
- Right-to-left languages layout
- Multi-tasking: split view and slide over

-->

adalpari committed:
Subscribing to the VM

adalpari committed:
OPening the webview

adalpari committed:
Some refactor and consuming url

adalpari committed:
Using the proper URL

adalpari committed:
Adding tests

adalpari committed:
style and detekt

adalpari commented on libs/analytics/src/main/java/org/wordpress/android/analytics/AnalyticsTracker.java:
```diff
@@ -1154,9 +1154,7 @@ public enum Stat {
-        VOICE_TO_CONTENT_BUTTON_RECORDING_LIMIT_REACHED,
-        BACKGROUND_REST_AUTODISCOVERY_SUCCESSFUL,
-        BACKGROUND_REST_AUTODISCOVERY_FAILED;
+        VOICE_TO_CONTENT_BUTTON_RECORDING_LIMIT_REACHED;
```
Waiting to see if this is useful anymore: https://a8c.slack.com/archives/C04PWEZSYFL/p1746192909267919

adalpari committed:
Readding backgrund check and event

nbradbury commented:
@adalpari Just adding this in case anyone else looks at this PR:

> comment the XML RPC lines as showing in the following screenshots (231, 232, 240)

I think you meant **340**, not **240**.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/LoginActivity.java:
```diff
@@ -709,6 +709,25 @@ public void gotXmlRpcEndpoint(String inputSiteAddress, String endpointAddress) {
     }
 
+    @Override public void gotRestEndpoint(String inputSiteAddress, String endpointAddress) {
+        CustomTabsIntent intent = new CustomTabsIntent.Builder()
```
Nitpick: Since we already use a `CustomTabsIntent` in `showWPcomLoginScreen`, would it make sense to add a function to show a url? ie: `showCustomTabsIntent(uri: Uri)`

adalpari replied:
Good catch!

nbradbury commented on gradle/libs.versions.toml:
```diff
@@ -107,6 +107,7 @@ wordpress-rs = 'alpha-20250411'
 wordpress-utils = '3.14.0'
 automattic-ucrop = '2.2.11'
 zendesk = '5.4.0'
+turbine = '1.2.0'
```
@adalpari I've not heard of Turbine before, but it looks useful. Was this something you used for Tumblr?

adalpari replied:
Yup. It's very useful to test coroutines and `Flow` https://github.com/cashapp/turbine

nbradbury reviewed (commented):
@adalpari This looks and works fine but unfortunately there are failing checks that need to be resolved.

adalpari committed:
Extracting CustomTabs creation

adalpari committed:
Workaround for lint crash

The link is crashing (not showing an error) because of some libraries incompatibility. The problem is that updating them is not trivial, as it involves upgrading the AGP as well. So, let's go with this workaround.
After internal conversation, next step is to change a bit the logic. So, hopefully flow will be used in a different way and lint wont crash.

adalpari commented:
> @adalpari This looks and works fine but unfortunately there are failing checks that need to be resolved.

Tricky one. Did a workaround but I don0t like it and I'm reverting it :(

adalpari committed:
Reverting workaround

adalpari commented:
Since some code needed to be moved because after a [slack](https://a8c.slack.com/archives/C04PWEZSYFL/p1746198417041819) discussion, I'm closing this PR in favor of this other one: https://github.com/wordpress-mobile/WordPress-Android/pull/21855
Similar code, different place.
